GoReleaser | added GitHub release workflows.

RutvikS-crest opened the pull request:
## Description

- This workflow will release a new Terraform Provider version to Hashicorp registry when a new tag in vx.x.x format is pushed to the repository.
- added `GPG_PRIVATE_KEY` and `PASSPHRASE` repository secrets with dummy values. we need to create and update respective secrets with a valid [Signing key](https://developer.hashicorp.com/terraform/registry/providers/publishing#preparing-and-adding-a-signing-key).

## Type of Change

- [ ] Bug Fix
- [x] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

<!-- TODO: Update the link below to point to your project's contributing guidelines -->
- [x] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

RutvikS-crest commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,51 @@
+# You will need to pass the `--batch` flag to `gpg` in your signing step 
+# in `goreleaser` to indicate this is being used in a non-interactive mode.
+#
+name: release
```
This workflow will release a new Terraform Provider version to Hashicorp registry when a new tag in vx.x.x format is pushed to the repository.

cisco-service commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,51 @@
+        env:
+          GPG_FINGERPRINT: ${{ steps.import_gpg.outputs.fingerprint }}
+          # GitHub sets this automatically
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[NIT] missing newline

lelia replied:
[NIT] missing newline

lelia commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,51 @@
+        uses: crazy-max/ghaction-import-gpg@111c56156bcc6918c056dbef52164cfa583dc549 # v5.2.0
+        id: import_gpg
+        with:
+          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          passphrase: ${{ secrets.PASSPHRASE }}
```
@RutvikS-crest I've added the GPG key and passphrase as GitHub secrets, matching the above values. Can you confirm whether this workflow requires any further modification or testing, now that the keys are in place?

RutvikS-crest replied:
That's it from repo side, I am assuming the key is also registered with Cisco/DevNet space in Hashicorp registry.

rquedas replied:
@lelia , can you confirm Rutvik's assumption is correct?

lelia replied:
@rquedas @RutvikS-crest it should now be registered both with Hashicorp and GitHub, yes. There is a `cisco-service` account (shared between CiscoDevNet and a few other Cisco open source orgs) which we have used to upload the new keys to.

lelia replied:
@RutvikS-crest FYI that I'm going to remove the repository-level placeholders for `GPG_PRIVATE_KEY` and `PASSPHRASE`, as they are currently superseding the organization-level secrets that I've added.

lelia commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,108 @@
+  push:
+    paths-ignore:
+      - 'README.md'
+  # For systems with an upstream API that could drift unexpectedly (like most SaaS systems, etc.),
+  # we recommend testing at a regular interval not necessarily tied to code changes. This will 
+  # ensure you are alerted to something breaking due to an API change, even if the code did not
+  # change.
+  # schedule:
+  #   - cron: '0 13 * * *'
```
Should we remove this?

RutvikS-crest replied:
Yep, Removed it

lelia commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,108 @@
+      env:
+        TF_ACC: "1"
+
+        # Set whatever additional acceptance test env vars here. You can
+        # optionally use data from your repository secrets using the
+        # following syntax:
+        # SOME_VAR: ${{ secrets.SOME_VAR }}
```
Do we need this?

lelia commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,108 @@
+        # SOME_VAR: ${{ secrets.SOME_VAR }}
+
+      run: |
+        go test -v -cover ./internal/provider/
```
[NIT] missing newline

lelia reviewed (changes requested):
@RutvikS-crest thanks for putting this together! Some minor NITs and questions about further verification now that the repository has access to the signing keys.

lelia reviewed (changes requested):
@RutvikS-crest these failing tests may point to a potential problem: https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35

Looks like it may still be referencing the original urlpath for the Go Client:

```bash
Run go mod download
go: github.com/cisco-open/appd-cloud-go-client@v1.0.2-0.202212280[4](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35#step:4:5)3829-8b9388d7[5](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5505021725/jobs/10031873219?pr=35#step:4:6)433: parsing go.mod:
	module declares its path as: github.com/aniketk-crest/appdynamicscloud-go-client
	        but was required as: github.com/cisco-open/appd-cloud-go-client
Error: Process completed with exit code 1.
```

FYI, you should have maintainer access to https://github.com/cisco-open/appd-cloud-go-client in case any further action is needed on the clientside to fix.

lelia commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,52 @@
+        uses: crazy-max/ghaction-import-gpg@111c56156bcc6918c056dbef52164cfa583dc549 # v5.2.0
+        id: import_gpg
+        with:
+          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          passphrase: ${{ secrets.PASSPHRASE }}
```
@RutvikS-crest @rquedas FYI that we've rotated the signing keys to use a more descriptive name, as they're now stored at the organizational level. Please accept my commit suggestions so the workflow will use the correct values, thanks!

```suggestion
          gpg_private_key: ${{ secrets.TERRAFORM_GPG_PRIVATE_KEY }}
          passphrase: ${{ secrets.TERRAFORM_GPG_PASSPHRASE }}
```

lelia reviewed (changes requested):
@RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?

RutvikS-crest commented:
> @RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?

@lelia, Tests are failing because they need access to AWS, Azure environments, I have added necessary environment variables in settings. You need to fill those in order to run tests. I have ran tests locally in my account it's running fine.

For now I have removed Workflows related to tests as they are optional to release the provider.

lelia commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,52 @@
+          GPG_FINGERPRINT: ${{ steps.import_gpg.outputs.fingerprint }}
+          # GitHub sets this automatically
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          
```
[NIT] missing newline

lelia commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,52 @@
+# This GitHub action can publish assets for release when a tag is created.
+# Currently its setup to run on any tag that matches the pattern "v*" (ie. v0.1.0).
+#
+# This uses an action (crazy-max/ghaction-import-gpg) that assumes you set your
+# private key in the `GPG_PRIVATE_KEY` secret and passphrase in the `PASSPHRASE`
+# secret. If you would rather own your own GPG handling, please fork this action
+# or use an alternative one for key handling.
```
```suggestion
# This uses an action (crazy-max/ghaction-import-gpg) that assumes you set your
# private key in the `TERRAFORM_GPG_PRIVATE_KEY` secret and passphrase in the
# `TERRAFORM_GPG_PASSPHRASE` secret. If you would rather own your own GPG handling,
# please fork this action or use an alternative one for key handling.
```

lelia commented:
> > @RutvikS-crest any theories on what's causing the [issues](https://github.com/cisco-open/terraform-provider-appd/actions/runs/5520871843?pr=35) with `go generate` and acceptance test jobs?
> 
> @lelia, Tests are failing because they need access to AWS, Azure environments, I have added necessary environment variables in settings. You need to fill those in order to run tests. I have ran tests locally in my account it's running fine.
> 
> For now I have removed Workflows related to tests as they are optional to release the provider.

@RutvikS-crest Thanks for the additional context! Could you [file an issue](https://github.com/cisco-open/terraform-provider-appd/issues/new/choose) describing what's currently failing, what needs to be done to fix the tests, and which specific environment variables need to be configured?

@rquedas After we merge this PR, can you work with @RutvikS-crest on resolving the above testing issue once we've successfully released on GitHub / published to the Terraform Registry?

lelia reviewed (changes requested):
Thanks for updating the GPG signing key names, @RutvikS-crest! 

Just a couple minor workflow fixes here and I think we'll be good to go.

lelia reviewed (approved):
Latest changes LGTM. @RutvikS-crest can you please file an issue as described in [my comment above](https://github.com/cisco-open/terraform-provider-appd/pull/35#issuecomment-1648681190) before we merge this? Thanks!

cginsberg98 commented:
@lelia and @RutvikS-crest  I am trying to report on the status but seems latest thread was 2 weeks ago, is the action with @RutvikS-crest  to close this out as per message above?
